pymc/logprob/censoring.py

```diff
@@ -237,3 +241,147 @@ def round_logprob(op, values, base_rv, **kwargs):
+    ]
+
+    # Maximum one branch allowed to be measurable
+    if len(measurable_comp_list) > 1:
+        return None
+
+    # If at least one of the branches is measurable
+    if len(measurable_comp_list) == 1:
+        measurable_comp_idx = measurable_comp_list[0]
+        measurable_component = components[measurable_comp_idx]
+
+        # broadcasting of the measurable component is not supported
+        if (
+            (measurable_component.type.broadcastable != node.outputs[0].broadcastable)
+            or (not compare_measurability_source([switch_condn, measurable_component], valued_rvs))
+            or (not rv_map_feature.request_measurable([measurable_component]))
+        ):
+            return None
+
+        if measurable_comp_idx == 0:
+            # changing the first branch of switch to always be the encoding
+            inverted_switch = pt.invert(switch_condn)
+
+            bitwise_op = MeasurableBitwise(inverted_switch.owner.op.scalar_op)
+            measurable_inverted_switch = bitwise_op.make_node(switch_condn).default_output()
+            encoded_rv = measurable_switch_encoding.make_node(
+                measurable_inverted_switch, *components[::-1]
+            ).default_output()
+
+            return [encoded_rv]
+
+    encoded_rv = measurable_switch_encoding.make_node(switch_condn, *components).default_output()
+
+    return [encoded_rv]
+
+
+@_logprob.register(MeasurableSwitchEncoding)
+def switch_encoding_logprob(op, values, *inputs, **kwargs):
+    (value,) = values
+
+    switch_condn, *components = inputs
+
+    # Right now, this only works for switch with both encoding branches.
+    logprob = pt.switch(
+        pt.eq(value, components[0]),
+        _logprob_helper(switch_condn, pt.as_tensor(np.array(True)), **kwargs),
+        pt.switch(
+            pt.eq(value, components[1]),
+            _logprob_helper(switch_condn, pt.as_tensor(np.array(False))),
+            -np.inf,
+        ),
+    )
+
+    # TODO: Calculate logprob for switch with one measurable component If RV is discrete,
+    #  give preference over encoding.
+
+    return logprob
+
+
+measurable_ir_rewrites_db.register(
+    "find_measurable_switch_encoding", find_measurable_switch_encoding, "basic", "censoring"
+)
+
+
+def compare_measurability_source(
+    inputs: Tuple[TensorVariable], valued_rvs: Container[TensorVariable]
+) -> bool:
+    ancestor_var_set = set()
+
+    # retrieve the source of measurability for all elements in 'inputs' separately.
+    for inp in inputs:
+        for ancestor_var in walk_model(
+            [inp],
+            walk_past_rvs=False,
+            stop_at_vars=set(valued_rvs),
+        ):
+            if (
+                ancestor_var.owner
+                and isinstance(ancestor_var.owner.op, RandomVariable)
+                and ancestor_var not in valued_rvs
+            ):
+                ancestor_var_set.add(ancestor_var)
+
+    return len(ancestor_var_set) == 1
+
+
+def walk_model(
```

Comment:
Probably okay to have this here temporarily. We could modify the other `walk_model` in the codebase such that we can use it to find `RandomVariable`s like in your use case.

IIRC, we have already two `walk_model` functions... not sure why